include/unifex/async_generator.hpp

```diff
@@ -0,0 +1,374 @@
+    // transition back to the correct execution context if necessary.
+    else if constexpr (
+        unifex::tag_invocable<
+            unifex::tag_t<unifex::await_transform>,
+            decltype(*this),
+            Value> ||
+        unifex::detail::_awaitable<Value>) {
+      return unifex::with_scheduler_affinity(
+          *this,
+          unifex::await_transform(*this, static_cast<Value&&>(value)),
+          *this->sched_);
+    } else {
+      // Otherwise, we don't know how to await this type. Just return it and
+      // let the compiler issue a diagnostic.
+      return (Value &&) value;
+    }
+  }
+
+private:
+  void transform_schedule_sender_impl_(unifex::any_scheduler newSched) {
+    this->sched_ = std::move(newSched);
+  }
+
+  template <typename ScheduleSender>
+  decltype(auto) transform_schedule_sender_(ScheduleSender&& snd) {
+    // This sender is a scheduler provider. Get the scheduler. This
+    // get_scheduler call returns a reference to the scheduler stored within
+    // snd, which is an object whose lifetime spans a suspend point. So it's
+    // ok to build an any_scheduler_ref from it:
+    transform_schedule_sender_impl_(unifex::get_scheduler(snd));
+
+    // Return the inner sender, appropriately wrapped in an awaitable:
+    return unifex::await_transform(
+        *this, std::forward<ScheduleSender>(snd).base());
+  }
+
+  // Friends with access to private fields
+  friend class async_generator<T>;
+  friend class async_generator_yield_operation<T>;
+  friend class reschedule_receiver<Promise>;
+
+  inline static constexpr unifex::inline_scheduler _default_scheduler{};
+
+  std::optional<unifex::any_operation_state_for<reschedule_receiver<Promise>>>
+      rescheduleOpSt_;
+  std::optional<unifex::any_scheduler> sched_;
+  // Keep track of the consumer scheduler
+  unifex::any_scheduler consumerSched_{_default_scheduler};
+  std::exception_ptr exception_;
+  // In this case, this keeps the consumer coroutine + a done() continuation.
+  // it's needed for at_coroutine exit for now, but also whenw e handle stop
+  // requests.
+  unifex::continuation_handle<> continuation_;
+  std::optional<any_receiver_ref<T&>> receiverOpt_;
+};
+
+}  // namespace detail
+
+template <typename T>
+class [[nodiscard]] async_generator {
+public:
+  using promise_type = detail::async_generator_promise<T>;
+
+  async_generator() noexcept : coroutine_(nullptr) {}
+
+  explicit async_generator(promise_type& promise) noexcept
+    : coroutine_(
+          unifex::coro::coroutine_handle<promise_type>::from_promise(promise)) {
+  }
+
+  async_generator(async_generator&& other) noexcept
+    : coroutine_(other.coroutine_) {
+    other.coroutine_ = nullptr;
+  }
+
+  ~async_generator() {
+    if (coroutine_) {
+      coroutine_.destroy();
+    }
+  }
+
+  async_generator& operator=(async_generator&& other) noexcept {
+    async_generator temp(std::move(other));
+    swap(temp);
```

Comment:
Eric Niebler taught me that you can do this in less code by receiving the argument by value.
```suggestion
  async_generator& operator=(async_generator other) noexcept {
    swap(other);
```